### PR TITLE
Minor: Add more documentation about Partitioning

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A
## Rationale for this change

@comphead  pointed out that the partitioning concept may not be well documented:  https://github.com/apache/arrow-datafusion/pull/8019/files#r1379167092

## What changes are included in this PR?

Try and document `Partitioning` better

## Are these changes tested?
N/A

## Are there any user-facing changes?
Docs

### CODE REVIEW

**Comment by alamb** [comment]
Maybe @andygrove  or someone knows  of a better source of documentation for Partitioning

File: datafusion/physical-expr/src/partitioning.rs [hunk 1]
```diff
@@ -15,14 +15,43 @@
...
 use std::sync::Arc;
 
 use crate::{expr_list_eq_strict_order, EquivalenceProperties, PhysicalExpr};
 
-/// Partitioning schemes supported by operators.
+/// Partitioning schemes supported by [`ExecutionPlan`]s.
+///
+/// A partition represents an independent stream that an `ExecutionPlan` can
+/// produce in parallel. Each `ExecutionPlan` must produce at least one
+/// partition, and the number of partitions varies based on the input and the
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:29]:
```suggestion
/// partition, and the number of output partitions varies based on the input and the
```

File: datafusion/physical-expr/src/partitioning.rs [hunk 2]
```diff
@@ -15,14 +15,43 @@
...
 use std::fmt;
 use std::sync::Arc;
 
 use crate::{expr_list_eq_strict_order, EquivalenceProperties, PhysicalExpr};
 
-/// Partitioning schemes supported by operators.
+/// Partitioning schemes supported by [`ExecutionPlan`]s.
+///
+/// A partition represents an independent stream that an `ExecutionPlan` can
+/// produce in parallel. Each `ExecutionPlan` must produce at least one
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:28]: Each `ExecutionPlan` instance must produce at least one partition, and consume at most one partition.

File: datafusion/physical-expr/src/partitioning.rs [hunk 3]
```diff
@@ -15,14 +15,43 @@
...
+/// file (note the actual DataFusion file scaners can read individual files in
+/// parallel, potentially producing multiple partitions per file)
+///
+/// Plans such as `SortPreservingMerge` produce a single output stream
+/// (1 output partition) by combining some number of input streams (input partitions)
+///
+/// Plans such as `FilterExec` produce the same number of output streams
+/// (partitions) as input streams (partitions).
+///
+/// [`ExecutionPlan`]: crate::physical_plan::ExecutionPlan
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:54]: we probably can add that every partition is the future by itself, and execution plan node is considered as completed when each partition(future) is completed

File: datafusion/physical-expr/src/partitioning.rs [hunk 4]
```diff
@@ -15,14 +15,43 @@
...
 
 use std::fmt;
 use std::sync::Arc;
 
 use crate::{expr_list_eq_strict_order, EquivalenceProperties, PhysicalExpr};
 
-/// Partitioning schemes supported by operators.
+/// Partitioning schemes supported by [`ExecutionPlan`]s.
+///
+/// A partition represents an independent stream that an `ExecutionPlan` can
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:27]:
```suggestion
/// A partition represents an independent stream of data batches that an `ExecutionPlan` can
```

File: datafusion/physical-expr/src/partitioning.rs [hunk 5]
```diff
@@ -15,14 +15,43 @@
...
+/// operation performed.
+///
+/// ```text
+///     ▲      ▲      ▲
+///     │      │      │
+///     │      │      │        An ExecutionPlan with 3
+///     │      │      │        output partitions will
+/// ┌───┴──────┴──────┴──┐     produce 3 streams of
+/// │   ExecutionPlan    │     RecordBatches that run in
+/// └────────────────────┘     parallel.
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:39]: would be nice to show the input partition as well

**Review by comphead — COMMENTED** [review]
Thanks @alamb for doing partition concepts easier to understand

File: datafusion/physical-expr/src/partitioning.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/partitioning.rs:28]: An ExecutionPlan can consume one or more partitions. I'll try and clarify that later today

File: datafusion/physical-expr/src/partitioning.rs (hunk 5 above)
Comment (alamb) [reply datafusion/physical-expr/src/partitioning.rs:39]: good idea -- I will do

File: datafusion/physical-expr/src/partitioning.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/partitioning.rs:54]: I would probably phrase it differently, like "the result of executing a Partition is a async stream (a kind of future)" or something. I'll try and clarify

File: datafusion/physical-expr/src/partitioning.rs (hunk 2 above)
Comment (comphead) [reply datafusion/physical-expr/src/partitioning.rs:28]: yes please my understanding by trait `.execute()` signature is to run the method for 1 partition at most
```
    fn execute(
        &self,
        partition: usize,
        context: Arc<TaskContext>,
    ) -> Result<SendableRecordBatchStream>;
```

File: datafusion/physical-expr/src/partitioning.rs [hunk 6]
```diff
@@ -15,14 +15,94 @@
...
+///                ▲                               ┌ ─ ─ ─ ─   ┌ ─ ─ ─ ─   ┌ ─ ─ ─ ─
+///                │                                 Input  │    Input  │    Input  │
+///                │                               │ Stream    │ Stream    │ Stream
+///                                                   (0)   │     (1)   │     (2)   │
+///               ...                              └ ─ ▲ ─ ─   └ ─ ▲ ─ ─   └ ─ ▲ ─ ─
+///                                                    │           │           │
+///                                                    │           │           │
+///                                                    │           │           │
+///
+// ExecutionPlan with 3 output                     3 (async) streams, one for each
```
Comment (comphead) [inline datafusion/physical-expr/src/partitioning.rs:61]: should we have 3 arrows in the diagram above?

Comment (alamb) [reply datafusion/physical-expr/src/partitioning.rs:61]: I am struggling with how to draw this -- the `ExecutionPlan` has a single child input (another `ExecutionPlan`) but both its child and itself produce 3 output partitions / streams

Comment (comphead) [reply datafusion/physical-expr/src/partitioning.rs:61]: wondering should we pursue this or not if its not feasible to draw?

Comment (alamb) [reply datafusion/physical-expr/src/partitioning.rs]: I purposely drew the execution plan with one arrow in and one arrow out to represent the fact it has one input child input `ExecutionPlan` and can be itself an input to another plan.  I tried to update the text to make this clearer.

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by alamb** [comment]
> lgtm thanks @alamb

Thank you so much for the reviews @comphead

**Pull request merged**